provisioner/hcp_sbom/provisioner.go

```diff
@@ -0,0 +1,251 @@
+	if downloadErr != nil {
+		return fmt.Errorf("failed to download file: %w", downloadErr)
+	}
+
+	// Download the file for user
+	p.downloadSBOMForUser(ui, comm)
+
+	// Validate the file
+	ui.Say(fmt.Sprintf("Validating SBOM file %s", destPath))
+	validationErr := p.validateSBOM(ui, destPath)
+	if validationErr != nil {
+		return fmt.Errorf("failed to validate SBOM file: %w", validationErr)
+	}
+
+	return nil
+}
+
+// downloadSBOMForPacker downloads SBOM from a specified source to a local
+// destination set by internal SBOM provisioner. It works with all communicators
+// from packersdk.
+func (p *Provisioner) downloadSBOMForPacker(
+	ui packersdk.Ui, comm packersdk.Communicator, generatedData map[string]interface{},
+) (string, error) {
+	src, err := interpolate.Render(p.config.Source, &p.config.ctx)
+	if err != nil {
+		return p.config.Destination, fmt.Errorf("error interpolating source: %s", err)
+	}
+
+	// FIXME:: Do we really need this?
+	// Check if the source is a JSON file
+	if filepath.Ext(src) != ".json" {
+		return p.config.Destination, fmt.Errorf(
+			"packer SBOM source file is not a JSON file: %s", src,
+		)
+	}
+
+	// Download the file for Packer
+	desti, ok := generatedData["dst"] // this has been set by HCPSBOMInternalProvisioner.Provision
+	if !ok {
+		return "", fmt.Errorf("failed to find location for Packer SBOM file")
+	}
+
+	dst := fmt.Sprintf("%v", desti)
+	// Ensure the destination directory exists
+	dir := filepath.Dir(dst)
+	if err := os.MkdirAll(dir, os.FileMode(0755)); err != nil {
+		return dst, fmt.Errorf("failed to create destination directory for Packer SBOM: %s", err)
+	}
+
+	// Open the destination file
+	f, err := os.OpenFile(dst, os.O_CREATE|os.O_WRONLY|os.O_TRUNC, 0644)
+	if err != nil {
+		return dst, fmt.Errorf("failed to open destination file: %s", err)
+	}
+	defer f.Close()
+
+	// Create MultiWriter for the current progress
+	pf := io.MultiWriter(f)
+
+	// Download the file
+	ui.Say(fmt.Sprintf("Downloading SBOM file %s for Packer => %s", src, dst))
+	if err = comm.Download(src, pf); err != nil {
+		ui.Error(fmt.Sprintf("download failed for Packer SBOM file: %s", err))
+		return dst, err
+	}
+
+	return dst, nil
+}
+
+// downloadSBOMForUser downloads a SBOM from a specified source to a local
+// destination given by user. It works with all communicators from packersdk.
+func (p *Provisioner) downloadSBOMForUser(
+	ui packersdk.Ui, comm packersdk.Communicator,
+) {
+	src, err := interpolate.Render(p.config.Source, &p.config.ctx)
+	if err != nil {
+		ui.Say(fmt.Sprintf("error interpolating source: %s", err))
+		return
+	}
+
+	// Determine the destination path
+	dst := p.config.Destination
+	if dst == "" {
+		ui.Say("skipped downloading SBOM file for user because 'Destination' is not provided")
+		return
+	}
+
+	dst, err = interpolate.Render(dst, &p.config.ctx)
+	if err != nil {
+		ui.Say(fmt.Sprintf("error interpolating SBOM file destination: %s", err))
+		return
+	}
+
+	if strings.HasSuffix(dst, "/") {
```

Comment:
This heuristic seems brittle, the output can be a directory even if the path doesn't end with `/`, we should rely on `os.Stat` regardless to know if this is a directory, irrespective of what the path looks like